File: 2019/python/day22.py
import math

CUT, DEAL_INC, DEAL_NEW = 0, 1, 2

def parse(l):
    if l.startswith('cut'):
        return (CUT, int(l[4:]))
    if l.startswith('deal with increment'):
        return (DEAL_INC, int(l[20:]))
    if l.startswith('deal into new stack'):
        return (DEAL_NEW, None)
 
def shuffle_position(pos, n_cards):
    for (s, n) in shuffle:
        if s == CUT:
            pos = (pos - n) % n_cards
        elif s == DEAL_NEW:
            pos = n_cards - 1 - pos
        elif s == DEAL_INC:
            pos = (pos * n) % n_cards
    return pos

def shuffle_position_rev(pos, n_cards):
    for (s, n) in reversed(shuffle):
        if s == CUT:
            pos = (pos + n) % n_cards
        elif s == DEAL_NEW:
            pos = n_cards - 1 - pos
        elif s == DEAL_INC:
            inv = pow(n, n_cards - 2, n_cards)
            pos = (pos * inv) % n_cards
    return pos

def compose_n_times(n, a, b, m):
    ra, rb = 1, 0
    while n != 0:
        if n & 1 == 1:
            ra, rb = (ra * a) % m, (ra * b + rb) % m
        n = n >> 1
        a, b = (a * a) % m, (a * b + b) % m
    return ra, rb

with open('../day_22.txt', 'r') as f:
    shuffle = [parse(l) for l in f]


cards = 10007
print('part 1:', shuffle_position(2019, cards))


cards      = 119315717514047
n_shuffles = 101741582076661

# shuffle(p) = (a * p + b) % cards 
b = shuffle_position_rev(0, cards)
a = shuffle_position_rev(1, cards) - b

an, bn = compose_n_times(n_shuffles, a, b, cards)

# suffle_n(p) = a^n * p + (a^(n-1) + a^(n-2) + ... + a^1 + 1) * b % cards
#             = a^n * p + (a^n - 1)/(a - 1) * b % cards 
# an = pow(a, n_shuffles, cards)
# bn = (an - 1) * pow(a - 1, cards - 2, cards) * b % cards

print('part 2:', (an * 2020 + bn) % cards)

File: 2019/go/day22.go
package main

import (
	"AdventOfCode/util"
	"fmt"
	"math/big"
	"strconv"
	"strings"
)

const (
	cut = iota
	dealNew
	dealInc
)

type shuffle struct {
	technique int
	num       int
}

func parse(lines []string) []shuffle {
	s := make([]shuffle, len(lines))
	for i, l := range lines {
		if strings.HasPrefix(l, "cut ") {
			n, _ := strconv.Atoi(l[4:])
			s[i].technique = cut
			s[i].num = n
		} else if strings.HasPrefix(l, "deal with increment ") {
			n, _ := strconv.Atoi(l[20:])
			s[i].technique = dealInc
			s[i].num = n
		} else if strings.HasPrefix(l, "deal into new stack") {
			s[i].technique = dealNew
		}
	}
	return s
}

func linearTransf(shuffles []shuffle, mod *big.Int) (*big.Int, *big.Int) {
	a, b := big.NewInt(1), big.NewInt(0)
	for _, s := range shuffles {
		num := big.NewInt(int64(s.num))
		switch s.technique {
		case cut:
			b.Sub(b, num).Mod(b, mod)
		case dealNew:
			a.Neg(a).Mod(a, mod)
			b.Sub(mod, b).Sub(b, big.NewInt(1))
		case dealInc:
			a.Mul(a, num).Mod(a, mod)
			b.Mul(b, num).Mod(b, mod)
		}
	}
	return a, b
}

func part1(shuffles []shuffle) int64 {
	const nCards = 10007
	mod := big.NewInt(int64(nCards))
	a, b := linearTransf(shuffles, mod)
	r := new(big.Int)
	r.Mul(a, big.NewInt(2019)).Add(r, b).Mod(r, mod)
	return r.Int64()
}

func part2(shuffles []shuffle) int64 {
	const (
		nCards = 119315717514047
		nTimes = 101741582076661
	)
	mod, exp, one := big.NewInt(int64(nCards)), big.NewInt(int64(nTimes)), big.NewInt(1)
	a, b := linearTransf(shuffles, mod)

	// inverse transformation
	a.ModInverse(a, mod)
	b.Neg(b).Mul(a, b).Mod(b, mod)

	// compose nTimes
	ca, cb := new(big.Int), new(big.Int)
	ca.Exp(a, exp, mod)
	cb.Sub(ca, one).Mul(cb, b)
	a.Sub(a, one).ModInverse(a, mod)
	cb.Mul(cb, a)

	r := new(big.Int)
	r.Mul(ca, big.NewInt(2020)).Add(r, cb).Mod(r, mod)
	return r.Int64()
}

func main() {
	input, _ := util.ReadInput("../day_22.txt")
	shuffles := parse(input)
	fmt.Printf("part 1: %d\n", part1(shuffles))
	fmt.Printf("part 2: %d\n", part2(shuffles))
}
